### ISSUE TITLE
[react-native] Cleanup type definitions

### ISSUE THREAD

**Issue by gyzerok** [DefinitelyTyped/DefinitelyTyped#15423]
TLDR;: This issue is intended to raise a discussion regarding possible reorganisation of the React Native type definitions.

All the following is mainly written for @alloy, however it's open for all for the discussion.

### Intro

After spending some time contributing to the React Native definitions I found myself thinking about following big changes that might benefit them. If we can come up to the agreement regarding some of this suggestions we can then split work into chunks and do it together. In other words let's introduce it as small PRs (one component at a time) rather then one huge.

### Remove docs comments

Currently most of the properties/functions definitions are accompanied with docs. As we know comments [always get stale](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/react-native/index.d.ts#L3367). And even though it might be useful to give a person nice description of what property is about on hover, it would be better to not show description at all then show the wrong one.

Since this repository is a peer effort on types it's harder to keep track if docs has changed or not especially when the release schedule for the library is every month. If for code people come to change something because it doesn't work anymore, I guess nobody or almost nobody comes here to change docs.

As for my experience I always check official docs to make sure that I will get up to date correct information, which makes comments in the definitions even more useless.

### Remove platform specific properties

Currently there are [platform](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/react-native/index.d.ts#L3230) [specific](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/react-native/index.d.ts#L3274) properties defined for each component which then got merged together with ["common"](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/react-native/index.d.ts#L3294) properties using extension mechanism.

First of all sometimes it seems to be impossible to split type definitions. For example for [union types](https://facebook.github.io/react-native/docs/image.html#resizemode). It means that we either duplicate definitions for iOS with all described alternatives or [put them under "common" properties](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/354cec620daccfa0ad167ba046651fb5fef69e8a/types/react-native/index.d.ts#L3341). In first case it may lead to errors when type is updated in one place, but not updated in another. In second case it means that we can't clearly separate platform specific definitions.

Also sometimes React Native adds support for both platforms, so some property should be moved form platform specific type to "common" type. But since everything work even without it, most of the time it doesn't happen.

I might be wrong, but I see no explicit value for consumers of type definitions in having platform specific properties types. Am I missing something?

### Rearrange fields in properties in accordance to docs

It would be much easier to update types for components if order of fields in will follow React Native official documentation. There is no way to enforce it, but we can at least make it so. And I guess it would quite easy for future contributors to spot this pattern and follow it.

**Comment by alloy** [comment]
Hey @gyzerok, thanks for the thoughtful message. I’m still finding my bearings in terms of maintaining these typings, so I don’t have a full clear thought on all of this yet, but I get where you’re coming from.

> ### Remove docs comments

I completely get what you’re saying regarding docs, but I’d like to give it some more thought first. E.g. what do you think about importing/updating docs with an automated tool instead? Afaik there’s no need to adjust the docs, so they could just be copied verbatim.

@orta I seem to recall you doing something like this for Jest, is that correct?

> ### Remove platform specific properties
> …
> For example for union types.

FYI You’re link regarding union types goes to a private Trello board.

> I might be wrong, but I see no explicit value for consumers of type definitions in having platform specific properties types. Am I missing something?

Maybe I’m misunderstanding what you are suggesting, but is it not possible to use the platform specific definitions and only get the ones that are actually supported? (Haven’t tried this yet myself.)

> ### Rearrange fields in properties in accordance to docs

Is there a specific sort (alphabetical) or is it random and do you just want to follow that exactly?

**Comment by orta** [comment]
With Jest I just did it all manually, they keep docs (API reference) in the website instead of inline, making it pretty tough to automate outside of parsing the HTML, then doing lookups against that. The API scope is smaller, so wasn't worth the automation. Plus it's maturing, less changes now.

However the React Native  API reference _is_ generated from source: e.g.

* http://facebook.github.io/react-native/docs/nativemethodsmixin.html
* https://github.com/facebook/react-native/blob/master/Libraries/Renderer/src/renderers/native/NativeMethodsMixin.js#L38

So it could be possible to generate those en-masse.

I can give it a shot if people are interested?

**Comment by gyzerok** [comment]
@orta yeah, I guess it would be the best option. Not sure how feasible it is. I'd even say that we either automate documentation alongside types or remove it.

**Comment by alloy** [comment]
@orta Please do 👍 

@gyzerok Yeah if automating won’t work, then I’m open to the idea of removing docs altogether.

**Comment by gyzerok** [comment]
> I completely get what you’re saying regarding docs, but I’d like to give it some more thought first. E.g. what do you think about importing/updating docs with an automated tool instead? Afaik there’s no need to adjust the docs, so they could just be copied verbatim.

Yeah it might be worth automation. However regardless of this there is not much sense in keeping current comments in type definitions. In other words it do not prevent up from cleaning comments right now.

> FYI You’re link regarding union types goes to a private Trello board.

Sorry, messed with the clipboard. Now link is correct.

> Maybe I’m misunderstanding what you are suggesting, but is it not possible to use the platform specific definitions and only get the ones that are actually supported? (Haven’t tried this yet myself.)

So I do not see the way how one would use them. Do you have any idea?

For example whenever you use `<Image />` you still have all the types available, because this is how it is defined right now.

> Is there a specific sort (alphabetical) or is it random and do you just want to follow that exactly?

It seems like it is alphabetical. One thing here that `android` and `ios` words seem to be part of signature as well. So first you will see all the common properties sorted alphabetically, then all the sorted Android props and then all the sorted iOS props.

**Comment by alloy** [comment]
> So I do not see the way how one would use them. Do you have any idea?
>
> For example whenever you use `<Image />` you still have all the types available, because this is how it is defined right now.

Yup yup, good food for thought 👍 I’ll also have a play with that to see if it’s usable in any way currently.

> It seems like it is alphabetical. One thing here that `android` and `ios` words seem to be part of signature as well. So first you will see all the common properties sorted alphabetically, then all the sorted Android props and then all the sorted iOS props.

Maybe we can have a (custom) linter rule for this 🤔

**Comment by orta** [comment]
Took a quick hour long dive into automating the docs, looks _pretty_ feasible - we want to get access to the JSON generated [in here](https://github.com/facebook/react-native/blob/master/website/server/extractDocs.js#L441-L462) for [these](https://github.com/facebook/react-native/blob/master/website/server/docsList.js) files. Then do a lookup in the TSD for the classes and properties.

**Comment by gyzerok** [comment]
It seems like this issue got stalled. I am closing it since I do not have any plans for contribution in the near future. Feel free to reopen if needed.
